src/Features/CSharp/Portable/RemoveUnnecessaryLambdaExpression/CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,279 @@
+                return;
+            }
+
+            var anonymousFunction = (AnonymousFunctionExpressionSyntax)context.Node;
+
+            // Syntax checks first.
+
+            // Don't simplify static lambdas.  The user made them explicitly static to make it clear it doesn't capture
+            // anything.  If we get rid of the lambda it won't be clear anymore if capturing/allocs are happening or not.
+            if (anonymousFunction.Modifiers.Any(SyntaxKind.StaticKeyword))
+                return;
+
+            if (!TryGetAnonymousFunctionInvocation(anonymousFunction, out var invocation, out var wasAwaited))
+                return;
+
+            // If we had an async function, but we didn't await the expression inside then we can't convert this. The
+            // underlying value was wrapped into a task, and that won't work if directly referencing the function.
+            if (wasAwaited != anonymousFunction.Modifiers.Any(SyntaxKind.AsyncKeyword))
+                return;
+
+            // We have to have an invocation in the lambda like `() => X()` or `() => expr.X()`.
+            var invokedExpression = invocation.Expression;
+            if (invokedExpression is not SimpleNameSyntax and not MemberAccessExpressionSyntax)
+                return;
+
+            // lambda and invocation have to agree on number of parameters.
+            var parameters = GetParameters(anonymousFunction);
+            if (parameters.Count != invocation.ArgumentList.Arguments.Count)
+                return;
+
+            // parameters must be passed 1:1 from lambda to invocation.
+            for (int i = 0, n = parameters.Count - 1; i < n; i++)
+            {
+                var parameter = parameters[i];
+                var argument = invocation.ArgumentList.Arguments[i];
+
+                if (argument.Expression is not IdentifierNameSyntax argumentIdentifier)
+                    return;
+
+                if (parameter.Identifier.ValueText != argumentIdentifier.Identifier.ValueText)
+                    return;
+            }
+
+            // if we have `() => new C().X()` then converting to `new C().X` very much changes the meaning.
+            if (MayHaveSideEffects(invokedExpression))
+                return;
+
+            // Looks like a reasonable candidate to simplify.  Now switch to semantics to check for sure.
+
+            if (CSharpSemanticFactsService.Instance.IsInExpressionTree(semanticModel, anonymousFunction, expressionType, cancellationToken))
+                return;
+
+            // If we have `object obj = x => Goo(x);` we don't want to simplify.  The compiler warns if you write
+            // `object obj = Goo;` because of the conversion to a non-delegate type.
+            var lambdaTypeInfo = semanticModel.GetTypeInfo(anonymousFunction, cancellationToken);
+            if (lambdaTypeInfo.ConvertedType == null || lambdaTypeInfo.ConvertedType.SpecialType is SpecialType.System_Object)
+                return;
```

Comment:
We can simplify with an explicit cast, but it's an extra work for the codefix. If we want to keep the codefix simple, I'd extend the comment to explicitly state this.

Reply:
sure.